hugr-core/src/hugr/linking.rs

```diff
@@ -0,0 +1,520 @@
+        if other.get_parent(sn) != Some(other.module_root()) {
+            return Err(NodeLinkingError::NotChildOfRoot(sn));
+        }
+        match dirv {
+            NodeLinkingDirective::Add { replace } => {
+                for &r in replace {
+                    if let Some(old_sn) = trns.replace.insert(r, sn) {
+                        return Err(NodeLinkingError::NodeMultiplyReplaced(r, old_sn, sn));
+                    }
+                }
+            }
+            NodeLinkingDirective::UseExisting(tn) => {
+                trns.use_existing.insert(sn, *tn);
+            }
+        }
+    }
+    Ok(trns)
+}
+
+fn link_by_node<SN: HugrNode, TGT: LinkHugr + ?Sized>(
+    hugr: &mut TGT,
+    transfers: Transfers<SN, TGT::Node>,
+    node_map: &mut HashMap<SN, TGT::Node>,
+) {
+    // Resolve `use_existing` first in case the existing node is also replaced by
+    // a new node (which we know will not be in RHS of any entry in `replace`).
+    for (sn, tn) in transfers.use_existing {
+        let copy = node_map.remove(&sn).unwrap();
+        // Because of `UseExisting` we avoided adding `sn`s descendants
+        debug_assert_eq!(hugr.children(copy).next(), None);
+        replace_static_src(hugr, copy, tn);
+    }
+    for (tn, sn) in transfers.replace {
+        let new_node = *node_map.get(&sn).unwrap();
+        replace_static_src(hugr, tn, new_node);
+    }
+}
+
+fn replace_static_src<H: HugrMut + ?Sized>(hugr: &mut H, old_src: H::Node, new_src: H::Node) {
+    let targets = hugr.all_linked_inputs(old_src).collect::<Vec<_>>();
+    for (target, inport) in targets {
+        let (src_node, outport) = hugr.single_linked_output(target, inport).unwrap();
+        debug_assert_eq!(src_node, old_src);
+        hugr.disconnect(target, inport);
+        hugr.connect(new_src, outport, target, inport);
+    }
+    hugr.remove_subtree(old_src);
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::HashMap;
+
+    use cool_asserts::assert_matches;
+    use itertools::Itertools;
+
+    use super::{LinkHugr, NodeLinkingDirective, NodeLinkingError};
+    use crate::builder::test::{dfg_calling_defn_decl, simple_dfg_hugr};
+    use crate::hugr::hugrmut::test::check_calls_defn_decl;
+    use crate::ops::{FuncDecl, OpTag, OpTrait, handle::NodeHandle};
+    use crate::{HugrView, hugr::HugrMut, types::Signature};
+
+    #[test]
+    fn test_insert_link_nodes_add() {
+        // Default (non-linking) methods...just for comparison
+        let (insert, _, _) = dfg_calling_defn_decl();
+
+        let mut h = simple_dfg_hugr();
+        h.insert_from_view(h.entrypoint(), &insert);
+        check_calls_defn_decl(&h, false, false);
+
+        let mut h = simple_dfg_hugr();
+        h.insert_hugr(h.entrypoint(), insert);
+        check_calls_defn_decl(&h, false, false);
+
+        // Specify which decls to transfer. No real "linking" here though.
+        for (call1, call2) in [(false, false), (false, true), (true, false), (true, true)] {
```

Comment:
here you could use rstest #[case] or #[values] as in the `test_insert_forest` case :)